webtransport/README.md

```diff
@@ -0,0 +1,60 @@
+
+## Addressing
+
+Webtransport multiaddresses are composed of a QUIC multiaddr, followed by `/webtransport` and a list of multihashes of the certificates that the server uses.
+Examples:
+* `/ip4/1.2.3.4/udp/443/quic/webtransport/`
+* `/ip6/fe80::1ff:fe23:4567:890a/udp/1234/quic/webtransport/<hash1><hash2><hash3>`
+
+If it is able to present a CA-signed certificate the list of certificate hashes SHOULD be empty. When using self-signed certificates, the server MUST include the hash(es) in the multiaddr.
+
+## WebTransport HTTP endpoint
+
+WebTransport needs a HTTPS URL to establish a WebTransport session, e.g. `https://example.com/webtransport`. As multiaddrs don't allow the encoding of URLs, this spec standardizes the endpoint. The HTTP endpoint of a libp2p WebTransport servers MUST be located at `/.well-known/libp2p-webtransport`.
+
+To allow future evolution of the way we run the libp2p handshake over WebTransport, we use a URL parameter. The handshake described in this document MUST be signaled by setting the `type` URL parameter to `multistream`.
+
+Example: The WebTransport URL of a WebTransport server advertising `/ip4/1.2.3.4/udp/1443/quic/webtransport/` would be `https://1.2.3.4:1443/.well-known/libp2p-webtransport?type=multistream`.
+
+## Security Handshake
+
+Unfortunately, the self-signed certificate doesn't allow the nodes to authenticate each others' peer IDs. It is therefore necessary to run an additional libp2p handshake on a newly established WebTransport connection.
+Once a WebTransport session is established, the clients opens a new stream and initiates a libp2p handshake (selecting the security protocol by running multistream, then performing the selected handshake).
+
+Note: Once we include the security protocol in the multiaddr (see https://github.com/libp2p/specs/pull/353), we will be able to shave off two (!!) round-trips here: Not running multistream saves one round trip. Furthermore, we'll be able to run the WebTransport handshake (i.e. the CONNECT request) in parallel with the cryptographic handshake, for example by transmitting the first handshake message as a URL parameter. The specifics of this is left to a future iteration of this spec.
+
+## Securing Streams
+
+All streams other than the stream used for the security handshake are protected using Salsa20. Two (symmetric) keys are derived from the master secrect established during the handshake, one for sending and for receiving on the stream.
```

Comment:
It's unfortunate that there's no concept of the client identity at the transport level, so that we don't need an additional wrapping of encryption.

We should see if that's an API that can be proposed at the spec level.

Reply:
We don't need that and we don't need double encryption. We can sufficiently identify the connection with the server certificate. Basically:

1. The client _knows_ it's talking to a server with a key that matches the certificate hash.
2. The client then sends a challenge of `(cert_hash, client_pid, server_pid, client_salt)`
3. The server responds with `server_salt, SIG(HASH(cert_hash || client_pid || server_pid || client_salt || server_salt))` to authenticate their end.
4. The client verifies this, then signs the same thing (maybe it signs the signature too?).

- In step 3, the server ties their peer ID to the connection.
- In step 4, the client ties _their_ peer ID to the connection.

Reply:
> 1. The client knows it's talking to a server with a key that matches the certificate hash.

That is correct iff it received the certificate hash in a signed peer record. Otherwise, the MITM could have generated a multiaddr containing its own cert hash.
Is that something we can require? Does js-libp2p support signed peer records?

> 2. The client then sends a challenge of (cert_hash, client_pid, server_pid, client_salt)

Note: Unfortunately, the client doesn't know which `cert_hash` was used by WebTransport. It only knows that _one_ of the hashes was considered valid, so it would have to send all of the cert hashes to the server.

> 3. The server responds with server_salt, SIG(HASH(cert_hash || client_pid || server_pid || client_salt || server_salt)) to authenticate their end.

It is crucial that the server validates that it owns _all_ cert hashes, otherwise a MITM attack is trivial. Furthermore (unless the certificate hash was signed in step 1), the client has to trust that the server actually performs this verification. Note that this kind of trust is not necessary when using TLS or Noise.

Here's a diagram of that handshake:
<img width="662" alt="image" src="https://user-images.githubusercontent.com/1478487/162625375-72a2fe28-655e-459b-9100-6b07476d40e9.png">

Reply:
@Kubuxu Would you mind taking a look at this proposal (the handshake that @Stebalien suggested in particular)?

Reply:
> That is correct iff it received the certificate hash in a signed peer record. Otherwise, the MITM could have generated a multiaddr containing its own cert hash.

It doesn't need to be in a signed peer record, the server signs (and, as you noted, validates) this in step 2.

> Note: Unfortunately, the client doesn't know which cert_hash was used by WebTransport. It only knows that one of the hashes was considered valid, so it would have to send all of the cert hashes to the server.

Makes sense.

> It is crucial that the server validates that it owns all cert hashes, otherwise a MITM attack is trivial. Furthermore (unless the certificate hash was signed in step 1), the client has to trust that the server actually performs this verification. Note that this kind of trust is not necessary when using TLS or Noise.

Yes, the server needs to check it. But it's definitely not the clients problem if the server doesn't check. The server could just as easily leak their key.

Reply:
> It doesn't need to be in a signed peer record, the server signs (and, as you noted, validates) this in step 2.

Fair enough. As long as the client trusts the server to perform this check, we're good. What makes me a little bit nervous is that the client on its own can't verify that the connection isn't MITM'ed, which is different from the security guarantees you get from a regular TLS handshake. We probably can live with it (and as you say, the server could leak the key anyway), but it is a somewhat weaker security guarantee.

Reply:
Let's set goals for this protocol, it needs to bind: the current channel and both PeerIDs.
TLS provides facilities for this by means of `ExportKeyingMaterial`. Result from it is unique per established channel, so even if someone were to try to MitM the session, `ExportKeyingMaterial` result will be different on either side.

I think, but I need to consider it more, we can just do:
Both sides do: `ekm = ExportKeyingMaterial("libp2p-auth")`, `ekm` is now shared secret on both sides, if they are using the same channel.
Then the Client sends the signature `Signature(ekm, ClientPrivKey)` and its pubkey. Server does the same, sends `Signature(ekm, ServerPrivKey)` and its pubkey.

Then to verify both sides verify that pubkeys match PeerIDs and they verify if the signature received is of `ekm`. The `ekm` is never transmitted over wire.

---

Questions is whether something like `ExportKeyingMaterial` is available on the browser end.

Reply:
PubKeys can also be exchanged out of band, IIRC libp2p has some other facilities for it.

Reply:
`ExportKeyingMaterial` is unavailable on the browser side.

Reply:
I think @Stebalien 's proposal can work. It is just there are a lot of details to work out that I don't feel comfortable deciding on (in essence we are rolling our own secure channel protocol, just without encryption).

An alternative to that proposal could be to reuse existing implementation of Noise secure channel to establish a secure session within WebTransport, for the purpose of co-authentication and authentication of the cert_hash. 
The proposal would be to start a libp2p-noise channel over WebCrypot and add `cert_hash` into the `Data` field of [HandshakePayload](https://github.com/libp2p/go-libp2p-noise/blob/95f0a223708e1bc01567dfbfe9ba7bab2723c781/pb/payload.pb.go#L25). An input of a predicate to assert onto the Data field would need to be added into libp2p-noise.

I think this is the lowest friction and risk solution.